File: codes/javascript/gemini/complete/task_3/Solution.js
const readline = require('readline');

function dispatchAmbulance(emergency_event, ambulances, hospitals, regions) {
    function calculateTravelTime(distance, regionType, congestionFactor) {
        let baseSpeed;
        switch (regionType) {
            case 'Urban':
                baseSpeed = 30;
                break;
            case 'Suburban':
                baseSpeed = 50;
                break;
            case 'Rural':
                baseSpeed = 80;
                break;
            default:
                baseSpeed = 50;
        }
        return Math.round((distance / (baseSpeed * congestionFactor)) * 60);
    }

    function buildGraph(regions) {
        const graph = {};
        regions.forEach(region => {
            graph[region.region_id] = {};
            for (const neighbor in region.distances) {
                if (!region.blocked_routes.includes(neighbor)) {
                    const travelTime = calculateTravelTime(
                        region.distances[neighbor],
                        region.type,
                        region.congestion_factor
                    );
                    graph[region.region_id][neighbor] = travelTime;
                } else {
                    graph[region.region_id][neighbor] = Infinity;
                }
            }
        });
        return graph;
    }

    function dijkstra(graph, start, end) {
        const distances = {};
        const visited = {};
        for (const node in graph) {
            distances[node] = Infinity;
        }
        distances[start] = 0;

        while (true) {
            let closestNode = null;
            for (const node in graph) {
                if (!visited[node] && (closestNode === null || distances[node] < distances[closestNode])) {
                    closestNode = node;
                }
            }

            if (closestNode === null) {
                break;
            }

            visited[closestNode] = true;

            for (const neighbor in graph[closestNode]) {
                const distance = graph[closestNode][neighbor];
                if (distances[closestNode] + distance < distances[neighbor]) {
                    distances[neighbor] = distances[closestNode] + distance;
                }
            }
        }

        return distances[end];
    }

    const graph = buildGraph(regions);
    const emergencyRegion = emergency_event.region_id;
    const requiredCapabilities = emergency_event.special_requirements;
    const isCritical = emergency_event.severity === 'Critical';

    let bestAmbulance = null;
    let minAmbulanceTime = Infinity;

    for (const ambulance of ambulances) {
        if (ambulance.status === 'Available') {
            let meetsRequirements = true;
            if (isCritical && !ambulance.capabilities.includes('Advanced Life Support')) {
                meetsRequirements = false;
            } else {
                for (const requirement of requiredCapabilities) {
                    if (!ambulance.capabilities.includes(requirement)) {
                        meetsRequirements = false;
                        break;
                    }
                }
            }


            if (meetsRequirements) {
                const ambulanceRegion = ambulance.current_region;
                const timeToEmergency = ambulanceRegion === emergencyRegion ? 0 : dijkstra(graph, ambulanceRegion, emergencyRegion);

                if (timeToEmergency < minAmbulanceTime) {
                    minAmbulanceTime = timeToEmergency;
                    bestAmbulance = ambulance;
                } else if (timeToEmergency === minAmbulanceTime) {
                    if (ambulance.capabilities.length > bestAmbulance.capabilities.length) {
                        bestAmbulance = ambulance;
                    } else if (ambulance.id < bestAmbulance.id) {
                        bestAmbulance = ambulance;
                    }
                }
            }
        }
    }

    if (!bestAmbulance) {
        return {
            emergency_id: emergency_event.id,
            assigned_ambulance: null,
            destination_hospital: null
        };
    }

    let bestHospital = null;
    let minHospitalTime = Infinity;

    for (const hospital of hospitals) {
        let meetsRequirements = true;

        if (emergency_event.type === 'Cardiac' && !hospital.capabilities.includes('Cardiac Center')) {
            meetsRequirements = false;
        }

        if (emergency_event.type === 'Trauma' && !hospital.capabilities.includes('Trauma Center')) {
            meetsRequirements = false;
        }

        if (meetsRequirements) {
            const hospitalRegion = hospital.region_id;
            const timeToHospital = emergencyRegion === hospitalRegion ? 0 : dijkstra(graph, emergencyRegion, hospitalRegion);

            if (timeToHospital < minHospitalTime) {
                minHospitalTime = timeToHospital;
                bestHospital = hospital;
            } else if (timeToHospital === minHospitalTime) {
                if (hospital.emergency_capacity === 'High' && bestHospital.emergency_capacity !== 'High') {
                    bestHospital = hospital;
                } else if (hospital.emergency_capacity === 'Medium' && bestHospital.emergency_capacity !== 'High' && bestHospital.emergency_capacity !== 'Medium') {
                    bestHospital = hospital;
                }
            }
        }
    }

    if (!bestHospital) {
        let nearestHospital = null;
        let minTime = Infinity;
        for (const hospital of hospitals) {
            const hospitalRegion = hospital.region_id;
            const timeToHospital = emergencyRegion === hospitalRegion ? 0 : dijkstra(graph, emergencyRegion, hospitalRegion);
            if (timeToHospital < minTime) {
                minTime = timeToHospital;
                nearestHospital = hospital;
            }
        }
        bestHospital = nearestHospital;
    }

    return {
        emergency_id: emergency_event.id,
        assigned_ambulance: bestAmbulance.id,
        destination_hospital: bestHospital.id
    };
}


if (require.main === module) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    let inputData = '';

    rl.on('line', (line) => {
        inputData += line;
    });

    rl.on('close', () => {
        try {
            const input = JSON.parse(inputData);
            const result = dispatchAmbulance(input.emergency_event, input.ambulances, input.hospitals, input.regions);
            console.log(JSON.stringify(result));
        } catch (error) {
            console.error('Error parsing JSON input:', error);
            console.log(JSON.stringify({
                emergency_id: null,
                assigned_ambulance: null,
                destination_hospital: null
            }));
        }
    });
}

File: codes/go/gemini/complete/task_3/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"sort"
)

func dispatchAmbulance(emergency_event map[string]interface{}, ambulances []interface{}, hospitals []interface{}, regions []interface{}) map[string]interface{} {
	result := map[string]interface{}{
		"emergency_id":         emergency_event["id"],
		"assigned_ambulance":   nil,
		"destination_hospital": nil,
	}

	emergencyRegionID := emergency_event["region_id"].(string)
	severity := emergency_event["severity"].(string)
	eventType := emergency_event["type"].(string)
	specialRequirements := emergency_event["special_requirements"].([]interface{})

	// --- Step 1: Select the fastest ambulance ---
	var bestAmbulance map[string]interface{}
	minAmbulanceTime := math.MaxFloat64

	for _, ambulanceInterface := range ambulances {
		ambulance := ambulanceInterface.(map[string]interface{})
		if ambulance["status"] == "Available" {
			ambulanceCapabilities := ambulance["capabilities"].([]interface{})
			currentRegion := ambulance["current_region"].(string)

			// Check if ambulance meets capability requirements
			meetsRequirements := true
			if severity == "Critical" && contains(specialRequirements, "Advanced Life Support") {
				hasALS := false
				for _, cap := range ambulanceCapabilities {
					if cap == "Advanced Life Support" {
						hasALS = true
						break
					}
				}
				if !hasALS {
					meetsRequirements = false
				}
			}

			if meetsRequirements {
				// Calculate response time
				responseTime := calculateShortestPath(currentRegion, emergencyRegionID, regions)

				if responseTime < minAmbulanceTime {
					minAmbulanceTime = responseTime
					bestAmbulance = ambulance
				} else if responseTime == minAmbulanceTime {
					// Prioritize by capabilities, then lower ID
					if len(ambulanceCapabilities) > len(bestAmbulance["capabilities"].([]interface{})) {
						bestAmbulance = ambulance
					} else if ambulance["id"].(string) < bestAmbulance["id"].(string) {
						bestAmbulance = ambulance
					}
				}
			}
		}
	}

	if bestAmbulance == nil {
		return result
	}

	result["assigned_ambulance"] = bestAmbulance["id"]

	// --- Step 2: Choose the best hospital ---
	var bestHospital map[string]interface{}
	minHospitalTime := math.MaxFloat64

	for _, hospitalInterface := range hospitals {
		hospital := hospitalInterface.(map[string]interface{})
		hospitalRegionID := hospital["region_id"].(string)
		hospitalCapabilities := hospital["capabilities"].([]interface{})
		emergencyCapacity := hospital["emergency_capacity"].(string)

		// Filter by matching capabilities
		meetsCapability := true
		if eventType == "Cardiac" {
			hasCardiacCenter := false
			for _, cap := range hospitalCapabilities {
				if cap == "Cardiac Center" {
					hasCardiacCenter = true
					break
				}
			}
			if !hasCardiacCenter && severity != "Critical" {
				meetsCapability = false
			}
		}

		if meetsCapability {
			// Calculate hospital arrival time
			hospitalArrivalTime := calculateShortestPath(emergencyRegionID, hospitalRegionID, regions)

			if hospitalArrivalTime < minHospitalTime {
				minHospitalTime = hospitalArrivalTime
				bestHospital = hospital
			} else if hospitalArrivalTime == minHospitalTime {
				// Prioritize by emergency capacity (High > Medium > Low)
				if emergencyCapacity == "High" && bestHospital["emergency_capacity"] != "High" {
					bestHospital = hospital
				} else if emergencyCapacity == "Medium" && bestHospital["emergency_capacity"] != "High" && bestHospital["emergency_capacity"] != "Medium" {
					bestHospital = hospital
				}
			}
		}
	}

	//If no suitable hospital: select nearest with emergency services
	if bestHospital == nil {
		minHospitalTime = math.MaxFloat64
		for _, hospitalInterface := range hospitals {
			hospital := hospitalInterface.(map[string]interface{})
			hospitalRegionID := hospital["region_id"].(string)
			hospitalArrivalTime := calculateShortestPath(emergencyRegionID, hospitalRegionID, regions)

			if hospitalArrivalTime < minHospitalTime {
				minHospitalTime = hospitalArrivalTime
				bestHospital = hospital
			}
		}
	}

	if bestHospital != nil {
		result["destination_hospital"] = bestHospital["id"]
	}

	return result
}

func calculateShortestPath(startRegionID string, endRegionID string, regions []interface{}) float64 {
	if startRegionID == endRegionID {
		return 0.0
	}

	graph := make(map[string]map[string]float64)
	regionMap := make(map[string]map[string]interface{})

	for _, regionInterface := range regions {
		region := regionInterface.(map[string]interface{})
		regionID := region["region_id"].(string)
		regionMap[regionID] = region

		graph[regionID] = make(map[string]float64)
		distances := region["distances"].(map[string]interface{})
		speed := region["speed"].(float64)
		congestionFactor := region["congestion_factor"].(float64)
		blockedRoutes := region["blocked_routes"].([]interface{})

		for neighborID, distanceInterface := range distances {
			distance := distanceInterface.(float64)

			isBlocked := false
			for _, blockedRouteInterface := range blockedRoutes {
				blockedRoute := blockedRouteInterface.(string)
				if blockedRoute == neighborID {
					isBlocked = true
					break
				}
			}

			if isBlocked {
				graph[regionID][neighborID] = math.MaxFloat64
			} else {
				baseSpeed := 0.0
				regionType := region["type"].(string)
				if regionType == "Urban" {
					baseSpeed = 30.0
				} else if regionType == "Suburban" {
					baseSpeed = 50.0
				} else {
					baseSpeed = 80.0
				}

				travelTime := distance / (baseSpeed * congestionFactor) * 60.0
				graph[regionID][neighborID] = travelTime
			}
		}
	}

	distances := make(map[string]float64)
	for regionID := range graph {
		distances[regionID] = math.MaxFloat64
	}
	distances[startRegionID] = 0.0

	visited := make(map[string]bool)

	for i := 0; i < len(graph); i++ {
		closestRegion := ""
		minDistance := math.MaxFloat64

		for regionID, distance := range distances {
			if !visited[regionID] && distance < minDistance {
				minDistance = distance
				closestRegion = regionID
			}
		}

		if closestRegion == "" {
			break
		}

		visited[closestRegion] = true

		for neighborID, weight := range graph[closestRegion] {
			if distances[closestRegion]+weight < distances[neighborID] {
				distances[neighborID] = distances[closestRegion] + weight
			}
		}
	}

	return math.Round(distances[endRegionID])
}

func contains(slice []interface{}, item interface{}) bool {
	for _, element := range slice {
		if element == item {
			return true
		}
	}
	return false
}

func main() {
	var input map[string]interface{}
	if err := json.NewDecoder(System.Stdin).Decode(&input); err != nil {
		panic(err)
	}

	emergency_event := input["emergency_event"].(map[string]interface{})
	ambulances := input["ambulances"].([]interface{})
	hospitals := input["hospitals"].([]interface{})
	regions := input["regions"].([]interface{})

	result := dispatchAmbulance(emergency_event, ambulances, hospitals, regions)

	jsonResult, err := json.Marshal(result)
	if err != nil {
		panic(err)
	}
	fmt.Println(string(jsonResult))
}